### PR TITLE
GraphQL: Make sure the context terminate

### LINKED ISSUE #26748: CVE-2022-2466 - Request Context not terminated with GraphQL
### Describe the bug

~~Upgrading from Quarkus 2.9.x to 2.10.x whenever i try to access `RoutingContext` request from a bean i get always the headers of the first request that have been made to the app.~~

~~This happens on both @GraphQLApi endpoints and event in any CDI bean that consumes `RoutingContext` or `HttpServerRequest`.~~

~~Also by debugging, the CurrentVertxRequest has always stale headers.~~

**Update:** The request context was not terminated. The issue is not related to the Routing Context.

My dependencies are the following:

```xml
   <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-kotlin</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-graphql</artifactId>
    </dependency>
    <dependency>
      <groupId>io.smallrye</groupId>
      <artifactId>smallrye-jwt</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-rest-client-jackson</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-arc</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-undertow</artifactId>
    </dependency>
    <dependency>
      <groupId>org.jetbrains.kotlin</groupId>
      <artifactId>kotlin-stdlib-jdk8</artifactId>
    </dependency>
```

CDI bean:

```kotlin
@ApplicationScoped
class JWTAwareContext @Inject constructor(
    val routingContext: RoutingContext
): Context {
    override fun getTenant(): String {
        return parseClaims()!!.getClaimValueAsString("realm")
    }

    override fun getUser(): String {
        return parseClaims()!!.getClaimValueAsString("email")
    }

    fun parseClaims(): JwtClaims? {
        val rawToken = this.routingContext.request().headers().get("Authorization")?.replace("Bearer ", "") // --> this will always stale in 2.10.x
        if (rawToken != null) {
            val jsonClaims = String(Base64.getUrlDecoder().decode(rawToken.split(".")[1]), StandardCharsets.UTF_8)
            return JwtClaims.parse(jsonClaims)
        }
        throw PassProNextForbiddenExceptionGQL(ApiError("Missing Token", ErrorCode.FORBIDDEN))
    }
}
```

Graphql endpoint

```kotlin

@GraphQLApi
class Test {
    @Inject lateinit var routingContext: RoutingContext


    @Query
    fun test(): String {
        return  "hello";
    }
}

```

With 2.9.x works correctly.

### Expected behavior

Everytime a new Request is performed by a client, the Request headers should be inline with the actual HTTP Request

### Actual behavior

With 2.10.x the first request headers became like cached value and any subsequent request headers will contain those instead of the actual headers

### How to Reproduce?

1) create an app with Quarkus 2.10.1 - 2.10.2 and the smallrye graphql extension
2) create an endpoint or a bean injecting RoutingContext
3) set some HTTP headers like Authorization, MyCustomHeader etc and send the http request
4) print RoutingContext.request().headers
5) set others HTTP headers  or remove the previous  and send the new http request
6) the second request headers will contain first request data also if you did not send them
7 ) switch to quarkus 2.9.x and will work as expected

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

adopt-openjdk-11.0.8

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.1 - 2.10.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

3.8.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by phillip-kruger**
Fix #26748
Added tests for this scenario.

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

### CODE REVIEW

**Review by sberyozkin — DISMISSED** [review]

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java [hunk 1]
```diff
@@ -58,9 +58,16 @@ public void handle(final RoutingContext ctx) {
...
+            ctx.response()
+                    .endHandler((e) -> {
+                        currentManagedContext.terminate();
+                    })
+                    .exceptionHandler((e) -> {
+                        currentManagedContext.terminate();
+                    })
+                    .closeHandler((e) -> {
+                        currentManagedContext.terminate();
+                    });
```
Comment (gsmet) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java:70]: Note that I'm not familiar at all with this. But I'm a bit worried you have to specifically declare these for every response.

Also what happens if `handleWithIdentity()` fails given you already have activated the context before calling the method?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java (hunk 1 above)
Comment (gsmet) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java:70]: Also, while we're at it, I would avoid creating 3 lambdas for this given it's runtime code.

**Comment by phillip-kruger** [comment]
@cescoffier if the method is in a try-finally, the context terminate too early.

**Comment by cescoffier** [comment]
Sorry, it should be in the throw clause not finally clause

**Comment by phillip-kruger** [comment]
You mean try - catch (Throwable) ?

**Review by gsmet — APPROVED** [review]
Looks good to me but I will let @cescoffier have the final say on it.

**Review by cescoffier — APPROVED** [review]

File: extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java [hunk 2]
```diff
@@ -56,11 +63,16 @@ public void handle(final RoutingContext ctx) {
...
-            ctx.response().bodyEndHandler((e) -> {
+            ctx.response()
+                    .endHandler(currentManagedContextTerminationHandler)
+                    .exceptionHandler(currentManagedContextTerminationHandler)
+                    .closeHandler(currentManagedContextTerminationHandler);
+
+            try {
+                handleWithIdentity(ctx);
+            } catch (Throwable t) {
                 currentManagedContext.terminate();
```
Comment (Ladicek) [inline extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java:74]: I'm not like totally exactly 1000% sure, but ... shouldn't we rethrow the exception? Or there's nothing above this handler that could meaningfully handle it and swallowing is fine?

Comment (cescoffier) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java:74]: You are totally right! 

@phillip-kruger can you open another PR?

Comment (gsmet) [reply extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLAbstractHandler.java:74]: I did something here as I would like it to be merged tonight: https://github.com/quarkusio/quarkus/pull/26797

**Pull request merged**